### PR TITLE
fix ordinal axis labels

### PR DESCRIPTION
**Pull request by EmmadUsmani**
Closes https://github.com/metabase/metabase/issues/38782
Closes https://github.com/metabase/metabase/issues/38783
Closes https://github.com/metabase/metabase/issues/39560

### Description

For ordinal x-axis labels, we should try to show every x-axis label. If the labels are too dense and no rotation setting was selected, then we automatically rotate the labels.

The implementation details of this fix are different than they were for the old dc.js implementation, since that implementation relied on measuring text boxes post-render, which we do not do with echarts. However, we try to closely match the old behavior, and in some cases improve it, see the demos below.

### Demo

https://github.com/metabase/metabase/assets/37751258/9d96253e-84e8-4759-9e42-902bc6108677

Comparing behavior between stats and echarts.

| Stats | ECharts |
| ----- | -------- |
| Screenshot 2024-03-26 at 3 41 49 PM | Screenshot 2024-03-27 at 2 29 52 PM |

https://github.com/metabase/metabase/assets/37751258/c8927ba6-3300-46b1-8218-1024faec29ca

Axis labels will be hidden if they take up ≥70% of the overall container height.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by EmmadUsmani** [comment]
* **#40693** <a href="https://app.graphite.dev/github/pr/metabase/metabase/40693?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `echarts`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EmmadUsmani and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts:198]: There is a TS error here

**Review by alxnddr — COMMENTED** [review]
I tried a few cases and noticed that behavior could be slightly improved to be similar to the previous implementation.

#### Label overlaps with ticks, rotation to 90 instead of 45
Screenshot 2024-03-28 at 1 18 43 AM

#### Missing ticks, incorrect spacing on sides

Summarize by Product -> Title. Filter: Product ID < 30

Stats:
Screenshot 2024-03-28 at 1 47 41 AM

Branch:
Screenshot 2024-03-28 at 1 47 19 AM

File: frontend/src/metabase/visualizations/echarts/cartesian/option/axis.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/cartesian/option/axis.ts:26]: We may want to store the computed rotation in the ChartMeasurements to avoid recomputation

File: frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts:198]: It's on the `echarts` branch, doesn't come from this PR.

File: frontend/src/metabase/visualizations/echarts/cartesian/option/axis.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/cartesian/option/axis.ts]: Done

**Comment by EmmadUsmani** [comment]
@alxnddr The behavior in your screenshots comes from `measureText` being incorrect. I don't want to fix it here because Anton is already making changes to it, there will be conflicts. Do you mind if I merge this PR without fixing that, and later we can improve the `measureText` function.

File: frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts:262]: I think we need to compute this before `firstXTickWidth` and  `lastXTickWidth`. For example, if `axisEnabledSetting` is auto (true) then we assume ticks are not rotated and the width can be substantial. So if we later change `axisEnabledSetting` to 90 degrees, it will produce incorrect paddings

**Comment by alxnddr** [comment]
It seems still we need to tweak the computation of `axisEnabledSetting` because even with sparse enough ticks it rotates ticks right away to 90 degrees. I applied the text measurement fix locally and got this:
Screenshot 2024-04-04 at 5 34 02 PM

**Comment by EmmadUsmani** [comment]
@alxnddr The ticks in that example aren't actually sparse, some of them are but the longer ones like November 2023, December 2023 will overlap. If we don't rotate in this case then echarts will automatically hide some dimension labels which does not look good, see the screenshot below.

Screenshot 2024-04-04 at 5 51 17 PM

I spent several hours today trying to fine tweak it some more, and sure it is possible to get it to use the 45 degree rotation here instead of 90 degree rotation, but then it looks bad on small dashcards where the 45 degree rotation is too crowded.

It's not possible to recreate the old behavior with 100% accuracy, the implementation details are different (old implementation used post-render overlap detection and measurement), and I don't think its worth spending a ton of time trying to do that if this behavior is good enough.

File: frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts:254]: Done

**Comment by alxnddr** [comment]
@EmmadUsmani yes, I wrote this regarding missing 45 degree auto rotation that I mentioned above. Even on a very sparse ticks it does not apply and I still think we need to fix this. I think it should be straightforward enough because we are not limited by ECharts API here and this is purely in our code. I can have a look

**Comment by alxnddr** [comment]
@EmmadUsmani done https://github.com/metabase/metabase/pull/41093

**Review by alxnddr — APPROVED** [review]
Looks great! Let's wait until the snapshots are updated and merge

**Pull request merged**
